sdks/python/apache_beam/transforms/environments.py

```diff
@@ -308,67 +309,80 @@ def from_options(cls, options):
-          'data_buffer_time_limit=%d' % self.data_buffer_time_limit)
+          'data_buffer_time_limit_ms=%d' % self.data_buffer_time_limit_ms)
     return 'EmbeddedPythonGrpcEnvironment(%s)' % ','.join(repr_parts)
 
   def to_runner_api_parameter(self, context):
-    if self.state_cache_size is None and self.data_buffer_time_limit is None:
-      payload = b''
-    elif self.state_cache_size is not None and \
-         self.data_buffer_time_limit is not None:
-      payload = b'%d,%d' % (self.state_cache_size, self.data_buffer_time_limit)
-    else:
-      # We want to make sure that the environment stays the same through the
-      # roundtrip to runner api, so here we don't want to set default for the
-      # other if only one of state cache size or data buffer time limit is set
-      raise ValueError(
-          'Must provide state cache size and data buffer time limit.')
+    params = {}
+    if self.state_cache_size is not None:
+      params['state_cache_size'] = self.state_cache_size
+    if self.data_buffer_time_limit_ms is not None:
+      params['data_buffer_time_limit_ms'] = self.data_buffer_time_limit_ms
+    payload = json.dumps(params).encode('utf-8')
     return python_urns.EMBEDDED_PYTHON_GRPC, payload
 
   @staticmethod
   def from_runner_api_parameter(payload, context):
     if payload:
-      state_cache_size, data_buffer_time_limit = \
-        payload.decode('utf-8').split(',')
+      state_cache_size, data_buffer_time_limit_ms = \
+        EmbeddedPythonGrpcEnvironment.parse_config(payload.decode('utf-8'))
       return EmbeddedPythonGrpcEnvironment(
-          state_cache_size=int(state_cache_size),
-          data_buffer_time_limit=int(data_buffer_time_limit))
+          state_cache_size=state_cache_size,
+          data_buffer_time_limit_ms=data_buffer_time_limit_ms)
     else:
       return EmbeddedPythonGrpcEnvironment()
 
   @classmethod
   def from_options(cls, options):
     if options.environment_config:
-      state_cache_size, data_buffer_time_limit = \
-        options.environment_config.split(',')
+      state_cache_size, data_buffer_time_limit_ms = \
+        EmbeddedPythonGrpcEnvironment.parse_config(options.environment_config)
       return cls(state_cache_size=state_cache_size,
-                 data_buffer_time_limit=data_buffer_time_limit)
+                 data_buffer_time_limit_ms=data_buffer_time_limit_ms)
     else:
       return cls()
 
+  @staticmethod
+  def parse_config(s):
+    if looks_like_json(s):
+      config = json.loads(s)
+      if 'state_cache_size' in config:
+        state_cache_size = int(config['state_cache_size'])
+      else:
+        state_cache_size = None
+
+      if 'data_buffer_time_limit_ms' in config:
+        data_buffer_time_limit_ms = int(config['data_buffer_time_limit_ms'])
+      else:
+        data_buffer_time_limit_ms = None
+    else:
+      state_cache_size = int(s)
+      data_buffer_time_limit_ms = None
+    return state_cache_size, data_buffer_time_limit_ms
```

Comment:
Would return a dict here because that will be more practical when we keep adding config entries.